src/compiler/checker.ts

```diff
@@ -16201,7 +16196,8 @@ namespace ts {
                 const constraint = getConstraintOfTypeParameter(inference.typeParameter);
                 if (constraint) {
                     const instantiatedConstraint = instantiateType(constraint, context.nonFixingMapper);
-                    if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {
+                    const isCompletionContext = contextFlags && (contextFlags & ContextFlags.Completion);
+                    if (!inferredType || isCompletionContext || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {

```

Comment:
This doesn't seem like the right place to do this - rather than passing the flag all the way thru signature resolution, if this flag is set, we should just look at the "base signature" (`getBaseSignature`) rather than starting inference at all, I think. Then again, if we did that we wouldn't get an instantiated constraint which reports more detailed information on circular constraints. Then again again, I'm pretty sure as-is this is caching into the `resolvedSignature` of a node, which this very much is not - which means requesting completions poisons the cache with incorrect types for later checking.

Reply:
Thanks for the feedback @weswigham! Do you have a suggestion how we still could fix the autocompletion for this? This was, unfortunately, the only way I found (with my limited knowledge of the codebase).

Maybe introducing a new method as an alternative to `getContextualType` just to be used by the completion service?

Reply:
I think `getContextualType` is still right, we just need to avoid using any signature caches.